datumaro/components/media.py

```diff
@@ -558,21 +605,109 @@ def __hash__(self):
         # Required for caching
         return hash((self._path, self._step, self._start_frame, self._end_frame))
 
+    @property
+    def path(self) -> str:
+        """Path to the media file"""
+        return self._path
+
+    @property
+    def ext(self) -> str:
+        """Media file extension (with the leading dot)"""
+        return osp.splitext(osp.basename(self.path))[1]
+
 
 class PointCloud(MediaElement):
     _type = MediaType.POINT_CLOUD
 
     def __init__(
         self,
-        path: str,
-        extra_images: Optional[List[Image]] = None,
+        extra_images: Union[List[Image], Callable[[], List[Image]], None] = None,
         crypter: Crypter = NULL_CRYPTER,
     ):
-        self._path = path
+        assert self.__class__ != PointCloud, (
+            f"Directly initalizing {self.__class__.__name__} is not supported. "
+            f"Please use fractory function '{self.__class__.__name__}.from_file()' "
+            f"or '{self.__class__.__name__}.from_data()'."
+        )
+        super().__init__(crypter)
+        self._extra_images = extra_images or []
 
-        self.extra_images: List[Image] = extra_images or []
+    @classmethod
+    def from_file(cls, path: str, *args, **kwargs):
+        return PointCloudFromFile(path, *args, **kwargs)
 
-        self._crypter = crypter
+    @classmethod
+    def from_data(cls, data: Union[bytes, Callable[[], bytes]], *args, **kwargs):
+        return PointCloudFromData(data, *args, **kwargs)
+
+    @property
+    def extra_images(self) -> List[Image]:
+        if callable(self._extra_images):
+            extra_images = self._extra_images()
+            assert isinstance(extra_images, list) and all(
+                [isinstance(image, Image) for image in extra_images]
+            )
+            return extra_images
+        return self._extra_images
+
+    def __eq__(self, other: object) -> bool:
+        return super().__eq__(other) and self.extra_images == other.extra_images
+
+
+class PointCloudFromFile(MediaElementFromFileMixin, PointCloud):
+    @property
+    def data(self) -> Optional[bytes]:
+        if os.path.exists(self.path):
+            with open(self.path, "rb") as f:
+                bytes_data = f.read()
+            return bytes_data
+        return None
+
+    def save(self, path, crypter: Crypter = NULL_CRYPTER):
+        if not crypter.is_null_crypter:
+            raise NotImplementedError(
+                f"{self.__class__.__name__} does not implement save() with non NullCrypter."
+            )
+
+        cur_path = osp.abspath(self.path) if self.path else None
+        path = osp.abspath(path)
+
+        if cur_path is not None and osp.isfile(cur_path):
+            if cur_path != path:
+                os.makedirs(osp.dirname(path), exist_ok=True)
+                shutil.copyfile(cur_path, path)
+        else:
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), cur_path)
```

Comment:
Not in this PR, but we could move saving extra images here in the future.

Reply:
Yes, I have a plan to save `extra_images` after refactoring `Image`.